crates/topos-p2p/src/behaviour/transmission/codec.rs

```diff
@@ -1,57 +1,107 @@
-use std::io;
+use std::{i8::MAX, io};
 
-use crate::constant::TRANSMISSION_PROTOCOL;
+use crate::constant::{SYNCHRONIZER_PROTOCOL, TRANSMISSION_PROTOCOL};
 
 use super::protocol::TransmissionProtocol;
-use futures::{AsyncRead, AsyncWrite, AsyncWriteExt};
+use futures::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
 use libp2p::{
     core::upgrade::{read_length_prefixed, write_length_prefixed},
     request_response::Codec,
     StreamProtocol,
 };
 
 #[derive(Clone)]
-pub(crate) struct TransmissionCodec();
+pub struct TransmissionCodec();
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct TransmissionRequest(pub(crate) Vec<u8>);
+pub enum TransmissionRequest {
+    Transmission(Vec<u8>),
+    Synchronizer(Vec<u8>),
+}
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct TransmissionResponse(pub(crate) Vec<u8>);
-
+pub enum TransmissionResponse {
+    Transmission(Vec<u8>),
+    Synchronizer(Vec<u8>),
+}
+const MAX_PACKET_SIZE: u64 = 16 * 1024 * 1024;
```

Comment:
I'm not sure I understand the connection between the difficulty of predicting the payload size and using `usize`. Can you elaborate? On most system where this code is likely to run, `usize` will be 64 bit but on a 32 bit cpu it will fail spectacularly, so why not settle that potential source for bugs right away and use `u64`?

Reply:
Even if we don't know what the max size of our payload is, our best estimate right now is exactly the code right now: 16777216 bytes. Wouldn't changing the max value technically be a hard fork of the chain? A node with max packet size set to something like 255bytes would not really be able to speak to anyone else. So not sure if making it a configurable parameter is plausible.

Reply:
> I'm not sure I understand the connection between the difficulty of predicting the payload size and using usize. Can you elaborate?

In the different methods that read the `io` we are dealing with a predefined buffer which receives the bytes read from the IO.
The buffer can only be initialized using `usize` as length arguments.
`Vec::with_capacity` or `vec![0; LENGTH]` are both accepting `usize`, not `u64` and that's why we're parsing the length (from the `io`) as `usize`  in order to read the exact length of the payload using `read_exact` which accept a mutable reference to a buffer.

> On most system where this code is likely to run, usize will be 64 bit but on a 32 bit cpu it will fail spectacularly, so why not settle that potential source for bugs right away and use u64?

**Spectacularly** is a bit strong, I think. In fact the codec will not fail, if a length of `u64::MAX` is encoded and received on a `32bits` arch the codec will use `usize` bound to `u32` and the parsing of the `length` will fail, the request will be denied with an `InvalidInput` error as it doesn't match something that the current system can handle.

And as you said, most systems are using 64bits, we can also define clearly that node needs to be run on 64bits arch. If a node admin runs it on a 32bits, the node will not be able to communicate on some channels, making it looks like a Byzantine node.